staging/src/k8s.io/kube-apiextensions-server/pkg/controller/finalizer/crd_finalizer.go

```diff
@@ -0,0 +1,297 @@
+		Reason:  "InstanceDeletionInProgress",
+		Message: "CustomResource deletion is in progress",
+	})
+	crd, err = c.crdClient.CustomResourceDefinitions().UpdateStatus(crd)
+	if err != nil {
+		return err
+	}
+
+	// Its possible for a naming conflict to have removed this resource from the API after instances were created.
+	// For now we will cowardly stop finalizing.  If we don't go through the REST API, weird things may happen:
+	// no audit trail, no admission checks or side effects, finalization would probably still work but defaulting
+	// would be missed.  It would be a mess.
+	// This requires human intervention to solve, update status so they have a reason.
+	if !apiextensions.IsCRDConditionFalse(crd, apiextensions.NameConflict) {
+		apiextensions.SetCRDCondition(crd, apiextensions.CustomResourceDefinitionCondition{
+			Type:    apiextensions.Terminating,
+			Status:  apiextensions.ConditionTrue,
+			Reason:  "InstanceDeletionStuck",
+			Message: fmt.Sprintf("cannot proceed with deletion because of %v condition", apiextensions.NameConflict),
+		})
+		crd, err = c.crdClient.CustomResourceDefinitions().UpdateStatus(crd)
+		if err != nil {
+			return err
+		}
+		return fmt.Errorf("cannot proceed with deletion because of %v condition", apiextensions.NameConflict)
+	}
+
+	// Now we can start deleting items.  We should use the REST API to ensure that all normal admission runs.
+	// Since we control the endpoints, we know that delete collection works.
+	crClient, err := c.clientPool.ClientForGroupVersionResource(schema.GroupVersionResource{Group: crd.Spec.Group, Version: crd.Spec.Version, Resource: crd.Status.AcceptedNames.Plural})
+	if err != nil {
+		return err
+	}
+	crAPIResource := &metav1.APIResource{
+		Name:         crd.Status.AcceptedNames.Plural,
+		SingularName: crd.Status.AcceptedNames.Singular,
+		Namespaced:   crd.Spec.Scope == apiextensions.NamespaceScoped,
+		Kind:         crd.Status.AcceptedNames.Kind,
+		Verbs:        metav1.Verbs([]string{"deletecollection", "list"}),
+		ShortNames:   crd.Status.AcceptedNames.ShortNames,
+	}
+	crResourceClient := crClient.Resource(crAPIResource, "" /* namespace all */)
+	allResources, err := crResourceClient.List(metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+
+	deletedNamespaces := sets.String{}
+	for _, item := range allResources.(*unstructured.UnstructuredList).Items {
+		metadata, err := meta.Accessor(&item)
+		if err != nil {
+			utilruntime.HandleError(err)
+			continue
+		}
+		if deletedNamespaces.Has(metadata.GetNamespace()) {
+			continue
+		}
+		if err := crClient.Resource(crAPIResource, metadata.GetNamespace()).DeleteCollection(nil, metav1.ListOptions{}); err != nil {
+			utilruntime.HandleError(err)
+			continue
+		}
+		deletedNamespaces.Insert(metadata.GetNamespace())
+	}
+
+	// now we need to wait until all the resources are deleted.  Start with a simple poll before we do anything fancy
+	listErr := wait.PollImmediate(1*time.Second, 5*time.Minute, func() (bool, error) {
```

Comment:
Each worker can be stuck up to 5 minutes? That could mean deletion of a CRD could be blocked through no fault of its own, just because other CRDs are slow to clean up? What if we just check if all items are gone, and if not we requeue (perhaps with a fixed initial delay)? That way other CRDs get a chance to clean up while we're waiting for this one.